### ISSUE TITLE
[feature request] Sum of several tensors

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#27522]
If I understand correctly `sum(tensor_list)` will allocate and keep O(N) intermediate tensors (same with a for loop) where N is number of tensors, which can be quite large in the case of big DenseNet. I propose to maybe generalize `torch.add` to support more than two tensors as input.

**Comment by pietern** [comment]
I agree that would be nice.

Would you be willing to submit a PR for this?

@gchanan What would this take on the dispatch side? I imagine this would require changing the core op to take N inputs, so it might be more complicated than it seems.

**Comment by SsnL** [comment]
Also wondering how hard it would be to automatically parallelize this with streams (when suited) in backend.

**Comment by vadimkantorov** [comment]
@pietern Unfortunately, I don't have enough time to figure out how to do it. I'm currently stuck even with a much simpler edit: https://github.com/pytorch/pytorch/pull/27621 ...

**Comment by gchanan** [comment]
do you mean `torch.sum(tensors)` or (python built-in) `sum(tensors)`?

**Comment by gchanan** [comment]
fwiw NumPy supports `np.sum(tuple of nd-arrays)`.

**Comment by fmassa** [comment]
@gchanan 
> fwiw NumPy supports np.sum(tuple of nd-arrays).

This is not equivalent to `sum(tuple of nd-arrays)` though.

In fact, numpy convert its input into an array in `np.sum` (and in pretty much all other methods), so `np.sum(tuple of nd-arrays)` becomes `np.sum(np.asarray(tuple of nd-arrays))`, which gives a number (sum over all elements).

But `sum(tuple of nd-arrays)` perform the sum of all the arrays, giving a `nd-array`.
```python
a = [np.random.rand(2, 2, 2) for _ in range(5)]
print(sum(a).shape)
# gives (2, 2, 2)
print(np.sum(a).shape)
# gives ()
```

TF has [a dedicated function for that](https://www.tensorflow.org/api_docs/python/tf/math/add_n).

The JIT can definitely optimize the particular pattern `sum(tuple of tensors)` into a single kernel, but I'm not sure if we should add a new function in PyTorch for that.

**Comment by vadimkantorov** [comment]
@fmasssa can we adapt `torch.add` to do this?

**Comment by vadimkantorov** [comment]
> do you mean `torch.sum(tensors)` or (python built-in) `sum(tensors)`?

I meant equivalent of `sum(tensors)`

**Comment by gchanan** [comment]
@vadimkantorov thanks for clarifying.

I _think_ all the dispatch building blocks are there for this, it's the same as `torch.cat` AFAICT.

**Comment by fmassa** [comment]
@vadimkantorov 

> can we adapt torch.add to do this?

we could, but `torch.add` already has two overloads, and I think adding support for a `tuple` / iterable to it might make it more confusing.

**Comment by vadimkantorov** [comment]
One way is to factor out the fused multipy-add into a separate method `torch.multiply_add` and then have `torch.add(*args)`

**Comment by vadimkantorov** [comment]
It would also be nice to have torch.logsumexp support multiple input tensors, so that prior torch.stack can be avoided. Actually, this kind of API makes sense for all reduction functions.

Alternatively, maybe this could be supported without API change if a NestedTensor becomes reality (so that torch.stack is done only logically without materialization).

**Comment by vadimkantorov** [comment]
In https://github.com/pytorch/pytorch/issues/32100 it seems that NumPy actually supports lists of tensors as reduction functions input instead of a stacked tensor

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/issues/38377

**Comment by vadimkantorov** [comment]
logaddexp applied to several tensors (without having to stack them and calling logsumexp) would also be useful to a CTC impl as above.

Given that https://github.com/pytorch/pytorch/pull/38384 is merged, I thought of bringing this up again: maybe via nested tensor.

Another instance that could be useful: torch.argmax across two tensors (without stacking them first). E.g. if we pass two tensors in that dimension, it would return for every element 0 or 1.

NumPy seems to support it for arbitrary functions via ufuncs: [`np.add.reduce([a,b])`](https://stackoverflow.com/posts/58091195/revisions)

**Comment by vadimkantorov** [comment]
Yay! NestedTensor seems to support reductions now: https://github.com/pytorch/nestedtensor/blob/master/examples/naryops_and_reduce.ipynb

some API musings:
```python
# if default nested dim is 0?
torch.add(NestedTensor([a, b]), dim = 0)

# maybe we can specify it semantically?
torch.add(NestedTensor([a, b]), dim = torch.nested_dimension)

# or have torch.stack return a nested tensor, since it's already an established way to bind tensors together, and new dim = 0 is expected
torch.add(torch.stack([a, b], layout = torch.nested), dim = 0)

# or maybe we can ask any op do torch.as_tensor and do torch.as_nested_tensor if it's a list of tensors
torch.add([a, b], dim = torch.nested_dimension)
```

@cpuhrsch

**Comment by mruberry** [comment]
@izdeby Maybe for each related?

**Comment by vadimkantorov** [comment]
Another in-the-wild example: https://github.com/microsoft/unilm/blob/b58d3ca6da7b8521b0f7bee7d9fcf6741d82233f/layoutlmft/layoutlmft/models/layoutlmv2/modeling_layoutlmv2.py#L659

@stas00

**Comment by stas00** [comment]
Are you suggesting that this is very inefficient memory allocation-wise and should be replaced with:
```
tensor_list = inputs_embeds + position_embeddings + spatial_position_embeddings + token_type_embeddings
functools.reduce(lambda acc, x: acc.add_(x), tensor_list, torch.zeros_like(tensor_list[0]))
```

until pytorch adds an efficient operator you proposed in this issue?

**Comment by vadimkantorov** [comment]
Yeah, regular addition of many tensors would result in N intermediate tensors if I understand well + I don't know if TorchScript optimization can optimize this.

So yeah, I propose to chime in this issue if you consider such reductions useful and in the meantime inplace addition is useful.

**Comment by vadimkantorov** [comment]
for me it's a bit surprising there's a peak of just one additional tensor size

**Comment by vadimkantorov** [comment]
Maybe if your test doesn't require grads they will be deallocated quickly

**Comment by stas00** [comment]
Yes, not O(N), but just O(2)

Feel free to experiment:

```
pip install ipyexperiments
```
and then 4 jupyter cells:
```
from ipyexperiments import IPyExperimentsPytorch
import torch
import functools
def consume_gpu_ram(n): return torch.ones((n, n)).cuda()
exp1 = IPyExperimentsPytorch()

t = [consume_gpu_ram(2**13) for x in range(4)]

y1 = t[0] + t[1] + t[2] + t[3]

y2 = functools.reduce(lambda acc, x: acc.add_(x), t, torch.zeros_like(t[0]))
```

**Comment by vadimkantorov** [comment]
From what I understand, in your test elements in `t` do not require gradients and thus this computation doesn't get tracked by autograd and are not kept around

**Comment by stas00** [comment]
If I use:
```
torch.ones((n, n), requires_grad=True)
```
then everything changes depending on the many2one function from which backward is called, e.g.:

![snapshot_24](https://user-images.githubusercontent.com/10676103/142797061-dce93f9e-29c9-4b91-854f-bfb1dd912834.png)
and:
![snapshot_25](https://user-images.githubusercontent.com/10676103/142797066-0d2282da-6905-4d91-a44d-b0df9c02aade.png)

note that in both cases, your version doesn't fair much different.

**Comment by stas00** [comment]
I suppose that's now what we normally do. We do this instead:

![snapshot_27](https://user-images.githubusercontent.com/10676103/142797867-9f633887-5296-404c-9f0f-17c1523c0721.png)

Adding more elements doesn't change the memory consumption:

![snapshot_28](https://user-images.githubusercontent.com/10676103/142798013-3d116374-bae3-4285-a00a-097edb2bb19b.png)

**Comment by cpuhrsch** [comment]
Is it at all realistic to use Triton here? Surely their tutorial on [summing vectors](https://triton-lang.org/getting-started/tutorials/01-vector-add.html#sphx-glr-getting-started-tutorials-01-vector-add-py) can be extended to multiple vectors. You'd then be left with implementing autograd for this operation, which can be achieved via [autograd.Function](https://pytorch.org/tutorials/beginner/examples_autograd/two_layer_net_custom_function.html).

**Comment by vadimkantorov** [comment]
I think it doens't make much sense to take a dependency on Triton just for this summation optimization. If it's used for all other multi-tensor reductions, it's more reasonable...

**Comment by cpuhrsch** [comment]
@vadimkantorov - You could guard against the presence of Trition and fall back to a less efficient implementation if it's not present. It should also allow you to unblock your perf explorations such as overall influence on the model and potential gains.

**Comment by vadimkantorov** [comment]
I think for summation specifically manual `functools.reduce(lambda acc, x: acc.add_(x), t, torch.zeros_like(t[0]))` works (if one copy for output is permissible)

**Comment by vadimkantorov** [comment]
But as https://github.com/pytorch/pytorch/issues/70533 mentions, probably the sum / functools.reduce workarounds aren't very TorchScript-able...

**Comment by cpuhrsch** [comment]
If you need TorchScript support you're unlikely to get around using the C++ extensibility features and ship your own binary code at this point. Neither Triton nor functools are JIT-able.

**Comment by vadimkantorov** [comment]
@cpuhrsch a real-world use-case for torch.sum(tensor_list, dim = 0): https://github.com/pytorch/pytorch/blob/master/torch/distributed/nn/functional.py#L298

also, could there be an improvement of using a single allocation instead of [torch.empty_like(tensor) for tensor in grad_outputs]? using nestedtensor (e.g. a _foreach_empty_like or a new general multi/foreach-allocation utility?) also if these grad_outputs are of the same size, a single empty() call could be used

**Comment by vadimkantorov** [comment]
another usecase to study about ops on multiple tensors and using or avoiding stack/copy:  https://github.com/pytorch/pytorch/commit/5654e6339879e438efb7cf50e88e356472eb0545#diff-8333640cff01dcc6a2ab20085003e8c998db905ad85c09603a20a6e363aff2e2R135

**Comment by vadimkantorov** [comment]
@cpuhrsch Also, one could imagine reduction across multiple tensors of different sizes without prior padding, e.g. using neutral elements as imputation: 0 for sum. Also might be possible for mean

**Comment by vadimkantorov** [comment]
@cpuhrsch Recent apple transformers optimization stresses reducing memory copies and also uses softmax on a tensorlist:

https://github.com/apple/ml-ane-transformers/blob/main/ane_transformers/huggingface/distilbert.py#L151

So supporting copy-avoiding TensorList (in addition to always-copying NestedTensor) at least for some ops may be a good idea

**Comment by cpuhrsch** [comment]
@vadimkantorov - while I agree, I'm wondering whether we couldn't achieve the same by mostly sticking to NestedTensors across the operations? As in, let's look at where the TensorList comes from to begin with and use a NestedTensor right away.

**Comment by vadimkantorov** [comment]
> @vadimkantorov - while I agree, I'm wondering whether we couldn't achieve the same by mostly sticking to NestedTensors across the operations? As in, let's look at where the TensorList comes from to begin with and use a NestedTensor right away.

I guess there would be a perf difference for re-allocation / copies that NestedTensor currently forces
